### ISSUE TITLE
AutoSelect Event is  Null when moved to 5.191.0 /CUSTCOM-179

### ISSUE THREAD

**Issue by bhanuunrivalled** [payara/Payara#4435]
# Description #
After upgrading to new Payara Version PrimeFaces AutoSelect in a Composite stops working 

## Expected Outcome

javax.faces.event.AjaxBehaviorEvent should not be Null

## Current Outcome

Currently we receive NULL Event

## Steps to reproduce (Only for bug reports) 
After deploying application
Open http://localhost:8080/javaeesimple-1.0-SNAPSHOT/autoCompleteTest.xhtml

This is working correctly in  Payara 4.1.2.174

### Samples

<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:h="http://xmlns.jcp.org/jsf/html"
      xmlns:p="http://primefaces.org/ui"
      xmlns:ui="http://xmlns.jcp.org/jsf/facelets"
      xmlns:cc="http://xmlns.jcp.org/jsf/composite"
      xmlns:f="http://xmlns.jcp.org/jsf/core">
<h:head>
    <title>component</title>
</h:head>

    <cc:interface>
        <cc:attribute name="id" required="true" />
        <cc:attribute name="matSelectListener" method-signature="void matSelectListener(javax.faces.event.AjaxBehaviorEvent)"/>
        <cc:attribute name="completeMatMethod" method-signature="java.util.List completeText(java.lang.String)"/>
        <cc:attribute name="matPerEntity" required="true"/>

    </cc:interface>

<cc:implementation>
    <h:form id="${cc.attrs.id}Form">
        <p:panelGrid id="${cc.attrs.id}Grid">
            <p:column>
                <h:outputText value="Material:" />
            </p:column>
            <p:autoComplete id="${cc.attrs.id}_single_material" queryDelay="3000" minQueryLength="5"
                            dropdown="true"  forceSelection="true"
                            completeMethod="#{cc.attrs.completeMatMethod}" value="#{cc.attrs.matPerEntity}">
                <p:ajax event="itemSelect" listener="#{cc.attrs.matSelectListener}" process="@this" update="@form"/>
            </p:autoComplete>
        </p:panelGrid>
    </h:form>
</cc:implementation>

</html>

##BEAN##

package com.achala.beans;

import org.omnifaces.cdi.ViewScoped;
import org.primefaces.event.SelectEvent;

import javax.faces.event.AjaxBehaviorEvent;
import javax.inject.Named;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

@Named
@ViewScoped
public class AutoCompleteView implements Serializable {

    private String txt6;

    public List<String> completeText(String query) {
        List<String> results = new ArrayList<String>();
        for(int i = 0; i < 10; i++) {
            results.add(query + i);
        }

        return results;
    }

    /**
     * Called when the user selects a material. Adds the selected material to the
     * list of selected materials
     *
     * @param e the event that carries the information about the selected material
     */
    public void onMatSelect(final AjaxBehaviorEvent e) {
        System.out.println("select event is ");
        System.out.println(e.toString());
    }

    /**
     * Creates a new material for the current PST
     */
    public void newMaterial() {
        System.out.println("action called");
    }

    public String getTxt6() {
        return txt6;
    }

    public void setTxt6(String txt6) {
        this.txt6 = txt6;
    }

}

[javaeesimple_autocompleteissue.zip](https://github.com/payara/Payara/files/4065563/javaeesimple_autocompleteissue.zip)

## Environment ##

- **Payara Version**: 5.191
- **Edition**: Full
- **JDK Version**:  1.8.0_131
- **Operating System**: Windows 10

**Comment by rdebusscher** [comment]
Hi @bhanuunrivalled 

This change seems due to a change in the WELD EL code.

This is the code of PrimeFaces which calls the listener. `org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl#processAjaxBehavior()``

        try {
            listener.invoke(elContext, new Object[]{});
        }
        catch (MethodNotFoundException mnfe) {
            processArgListener(context, elContext, event);
        }
        catch (IllegalArgumentException iae) {
            processArgListener(context, elContext, event);
        }

So first it calls the method without the event as parameters (`invoke(elContext, new Object[]{})`). In Payara 4.1.2.174, Weld throws an exception that the method isn't found. And then a second attempt is made with the parameter (by `processArgListener`)

In later versions (Payara 5.191, but also Payara 4.1.2.191) the Weld version accepts the first call and calls the Java method without any parameter value (hence the null)

It is not related to the usage of the custom component, it occurs also when PrimeFaces `p:autoComplete` is used in an facelets page directly.

It would be best that PrimeFaces would revert the calls, first with parameter and then without.

**Comment by rdebusscher** [comment]
I have to check the specifications on what should happen in the case PrimeFaces uses the listeners to determine who is to 'blame'

**Comment by bhanuunrivalled** [comment]
HI @rdebusscher ,

Thanks we debugged the same  org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl#processAjaxBehavior()

Even  we thought the issue could be with weld as replacing faces and primefaces resulted in the same issue .

**Comment by bhanuunrivalled** [comment]
HI @rdebusscher ,

We observed that it did not work with this also.

```
 <p:autoComplete id="${cc.attrs.id}_single_material" queryDelay="3000" minQueryLength="5"
                            dropdown="true"  forceSelection="true"
                            completeMethod="#{cc.attrs.completeMatMethod}" value="#{cc.attrs.matPerEntity}">
                <f:ajax event="itemSelect" listener="#{cc.attrs.matSelectListener}" execute="@this" render="@form" />

            </p:autoComplete>
```

**Comment by rdebusscher** [comment]
Hi,

I have created the internal ticket `CUSTCOM-179` to investigate the issue and some possible solutions.

Regards
Rudy

**Comment by dmatej** [comment]
New issue in PrimeFaces: https://github.com/primefaces/primefaces/issues/5827
Bug fix for PrimeFaces (tested with Payara): https://github.com/primefaces/primefaces/pull/5828

Until PrimeFaces fixes their code, this temporary workaround on Payara side will be required.

**Comment by rdebusscher** [comment]
Fix released as part of 5.2020.2
